feat(chat): Add tool calls and thinking/reasoning support to mo.ui.chat

adamwdraper opened the pull request:
## Summary

This PR adds comprehensive tool call and thinking/reasoning support to `mo.ui.chat`, with a new `mo.ai.llm.pydantic_ai` class that simplifies working with LLM tools.

![reasoning tools](https://github.com/user-attachments/assets/bab24529-7441-4a3a-8081-84afa834bc9f)


## Features

### 1. Tool Call Support in Chat UI
- Tool calls are displayed as collapsible accordions showing input/output
- Tools appear **before** the message bubble (same pattern as reasoning)
- Real-time streaming: tools show "calling" state, then update with results

### 2. Thinking/Reasoning Support
- Added `enable_thinking` parameter to `pydantic_ai` class
- Reasoning appears as a collapsible accordion before the response
- Streams in real-time as the LLM thinks

### 3. New `mo.ai.llm.pydantic_ai` Class
- Simplified tool integration using [pydantic-ai](https://ai.pydantic.dev/)
- Just pass Python functions as `tools` - pydantic-ai handles the rest
- Supports all major providers: OpenAI, Anthropic, Google, Groq, etc.

### 4. Multi-turn Conversation Support
- Properly maintains message history across turns
- Stores pydantic-ai's native messages to ensure correct `tool_use`/`tool_result` pairing (required by Claude)
- Unknown part types now pass through `ChatMessage` instead of being discarded

### 5. W&B Inference & OpenAI-Compatible API Support
- Added `base_url` parameter for connecting to OpenAI-compatible endpoints
- Added `api_key` parameter for direct credential passing (no env vars needed)
- Uses Pydantic AI Provider pattern for thread-safe, no-side-effect credential handling
- Automatic reasoning extraction from W&B Inference models via `OpenAIModelProfile`
- See [W&B Inference docs](https://docs.wandb.ai/inference/) for available models

## Example Usage

### With Anthropic (Tools + Thinking)
```python
import marimo as mo

def get_weather(location: str) -> dict:
    """Get current weather for a location."""
    return {"temperature": 72, "conditions": "sunny"}

chatbot = mo.ui.chat(
    mo.ai.llm.pydantic_ai(
        "anthropic:claude-sonnet-4-5",
        tools=[get_weather],
        enable_thinking=True,
        api_key=api_key,
    ),
)
```

### With W&B Inference (Reasoning Models)
```python
import marimo as mo

chatbot = mo.ui.chat(
    mo.ai.llm.pydantic_ai(
        "openai:deepseek-ai/DeepSeek-R1-0528",
        base_url="https://api.inference.wandb.ai/v1",
        api_key=wandb_api_key,
        enable_thinking=True,  # Extracts reasoning_content from response
    ),
)
```

## Files Changed

### Core Changes
- `marimo/_ai/llm/_impl.py` - New `pydantic_ai` class with streaming, tools, thinking, and W&B support
- `marimo/_ai/_types.py` - Allow unknown part types to pass through

### Frontend
- `frontend/src/plugins/impl/chat/chat-ui.tsx` - Render tool calls and reasoning accordions

### Examples
- `examples/ai/chat/pydantic_ai_with_thinking_and_tools.py` - Combined example with Anthropic
- `examples/ai/chat/wandb_inference_example.py` - W&B Inference with reasoning models

### Tests
- `tests/_ai/llm/test_impl.py` - Tests for thinking, history storage, Provider pattern, and W&B support

## Testing

- Tested multi-turn conversations with tools and thinking on Claude
- Tested W&B Inference with DeepSeek R1 reasoning model
- Verified tool calls display correctly during streaming
- Verified reasoning appears before response
- Added unit tests for new functionality

## Screenshots

The chat UI now shows:
1. **Reasoning accordion** (collapsible) - appears first as LLM thinks
2. **Tool call accordions** (collapsible) - shows tool name, inputs, and outputs
3. **Response text** - final synthesized answer

mscolnick commented on marimo/_ai/llm/_impl.py:
```diff
@@ -26,6 +36,187 @@
 )
 
 
+def _python_type_to_json_schema(py_type: Any) -> dict[str, Any]:
```
pydantic AI does this for you

adamwdraper commented:
@mscolnick addressed your comments:

1. Simplified pydantic_ai to accept both model string AND Agent
You can now use it two ways:
Simple (like other mo.ai.llm classes):
`mo.ai.llm.pydantic_ai(
    "anthropic:claude-sonnet-4-5",
    tools=[get_weather],
    instructions="You are helpful.",
    model_settings=AnthropicModelSettings(...),
)`
Existing pydantic ai user (pass a pre-configured Agent):
`from pydantic_ai import Agent

agent = Agent("anthropic:claude-sonnet-4-5", tools=[...], deps_type=MyDeps, ...)
mo.ai.llm.pydantic_ai(agent)
`

2. Removed internal complexity
❌ Removed _python_type_to_json_schema - Pydantic AI handles this
❌ Removed _create_model, _get_provider, _get_model_name - Agent creation delegated to Pydantic AI
❌ Removed _parse_docstring_args, _function_to_openai_tool - not needed

3. Future-proof via **kwargs
Any new Agent parameters Pydantic AI adds will work automatically without code changes.
The class is now a thin wrapper (~130 lines of logic) that just bridges Pydantic AI to marimo's chat UI.

mscolnick commented on marimo/_ai/llm/_impl.py:
```diff
@@ -664,3 +702,601 @@ def __call__(
+        self,
+        model: Any,
+        *,
+        tools: Optional[list[Callable[..., Any]]] = None,
```
having model and tools separate is still not ideal because that implies all providers/models can use tools (which is not true). allowing just an Agent object helps prevent this

mscolnick commented on frontend/src/plugins/impl/chat/chat-ui.tsx:
```diff
@@ -441,6 +555,28 @@ export const Chatbot: React.FC<Props> = (props) => {
             .join("\n");
 
+          // Separate tool parts, reasoning parts, and other parts for assistant messages
+          const toolParts =
```
we can share this with the chat sidebar `<RenderParts`. we may not have this now, but we can do some refactoring so that there is a generic component to render these parts and we don't need to unpack or adapt them

mscolnick commented on marimo/_ai/llm/_impl.py:
```diff
@@ -664,3 +702,411 @@ def __call__(
+        final_result: Any = None
+        pending_tool_calls: dict[str, dict[str, Any]] = {}
+
+        def _build_parts() -> list[dict[str, Any]]:
```
@Light2Dark is adding some features now that should be able to remove this. either we can merge this as-is and delete later or wait (just a few days)

Light2Dark commented:
Hi Adam, sorry this took a while, actively looking into this. I agree with some of the approaches and hopefully can push something up over the next few days.
